packages/labs/context/README.md

```diff
@@ -0,0 +1,101 @@
+# @lit-labs/context
+
+## Overview
+
+## Usage
+
+There are several different usages of the Context API.
+
+### Creating a Context
+
+First lets define a context key we can use elsewhere in our examples:
+
+```ts
+import {createContext} from '@lit-labs/context';
+
+export interface Logger {
+  log: (msg: string) => void;
+}
+
+export const loggerContext = createContext<Logger>('logger');
+```
+
+### Consuming a Context
+
+Now we can define a consumer for this context, some component in our app needs the logger:
+
+```ts
+import {ContextConsumer, property} from '@lit-labs/context';
+import {LitElement} from 'lit';
+import {Logger, loggerContext} from './logger.js';
+
+export class MyElement extends LitElement {
+  @property({attribute: false})
+  public logger?: Logger;
+
+  public constructor() {
+    // add the consumer controller
+    this.addController(
+      new ContextConsumer(
+        this,
+        loggerContext,
+        (value) => {
+          this.logger = value;
+        },
+        true // pass true to get updates if the logger changes
+      )
+    );
+  }
+}
+```
+
+Another way we can use a context in a component is via the `contextProvided` decorator:
+
+```ts
+import {ContextConsumer, property} from '@lit-labs/context';
+import {LitElement} from 'lit';
+import {Logger, loggerContext} from './logger.js';
+
+export class MyElement extends LitElement {
+  @contextProvided({context: loggerContext, multiple: true})
+  @property({attribute: false})
+  public logger?: Logger;
+}
+```
+
+### Providing a Context
+
+Finally we want to be able to provide this context from somewhere higher in the DOM:
+
+```ts
+import {LitElement} from 'lit';
+import '@lit-labs/context/lib/elements/context-provider-element.js';
+import {loggerContext, Logger} from './my-logger.js';
+
+const loggerContext = createContext('logger', new Logger());
+
+export class MyApp extends LitElement {
+  private logger: Logger = {
+    log: (msg) => {
+      console.log(`[my-app] ${msg}`);
+    },
+  };
+  protected render(): TemplateResult {
+    return html`
+      <context-provider .context=${loggerContext} .value=${this.logger}>
```

Comment:
I'd like to offer a way to provide context to a subtree w/o an element. A directive seems like a good way to do this, and can wrap the entire template too:

```ts
  protected render(): TemplateResult {
    return provide({context: loggerContext, value: this.logger}, html`
      <my-thing></my-thing>
    `;
  }
```

Although I'd probably use the controller if I'm providing to the whole tree. I also think a `@provide` decorator would be nice:

```ts
export class MyApp extends LitElement {
  @provide(loggerContext)
  // @state() - @provide() is non-reactive by default? Use @state() to make it reactive...
  private logger: Logger = {
    log: (msg) => {
      console.log(`[my-app] ${msg}`);
    },
  };
```

Reply:
Yeah I like both of those additions. I also then a generic function for just providing context at any `HTMLElement` outside of any lit-specific template or class context would be useful (we have a case where we could do with this right now):

```ts
const view = document.querySelector('#some-view');
provideContext(view, loggerContext, logger);
```

Reply:
We'd need to figure out when this would be called and what state it needs to preserve. If that function adds a listener each time, we don't want to call it in render/update and add multiple listeners... We can keep weakmaps/sets for that. This might be good as a method on ContextProvider?

Reply:
Trying to implement the directive, did your example above have an omission in it? I can't figure out how to make a directive that returns a `TemplateResult` directly? I'd have expected to have to wrap it in a `html` tag like so:

```ts
  protected render(): TemplateResult {
    return html`${provide({context: loggerContext, value: this.logger}, html`
      <my-thing></my-thing>
    `})`;
  }
  ```
  
  This tracks with the implementation I'm seeing in `until`.

Reply:
In the case of the ChildPart directive outlined here, would the directive attach the `context-request` event listener to the `parentNode` of the ChildPart or render a wrapper node around the content produced by the passed in template?

The `AttributePart` case is I think pretty straightforward, attach the listener to the `part.element` and remove it if the `part.element` changed?

Reply:
A DirectiveResult is a renderable value, so you don't need to wrap it. You can return a template from `render()` or `update()`.

For the ChildPart case there is a little complication about where to attach the listener, and to make sure that you only respond to events from the elements wrapped, so maybe that's not a good place to start.

I wouldn't do an AttributePart, I'd do ElementPart instead. I don't see a need for an attribute name.

Reply:
Ah I see. So we could skip support for Child part and just support ElementPart or direct return usage?

Reply:
I added a basic implementation of the `ElementPart` for `provide` directive. I am unsure how to extend this to use it as a direct returned templated result. Would it imply that the provide directive would render a passed in 3rd argument as the template? How do you get to the element from the template result inside the `update` function?